File: DAY 4/solutions.java/76. Minimum Window Substring.java
import java.util.*;

class Solution {
    public String minWindow(String s, String t) {

        if (t.length() > s.length()) {
            return "";
        }

        HashMap<Character, Integer> need = new HashMap<>();

        for (char ch : t.toCharArray()) {
            need.put(ch, need.getOrDefault(ch, 0) + 1);
        }

        HashMap<Character, Integer> window = new HashMap<>();

        int left = 0;
        int have = 0;
        int needCount = need.size();

        int minLength = Integer.MAX_VALUE;
        String answer = "";

        for (int right = 0; right < s.length(); right++) {

            char ch = s.charAt(right);

            window.put(ch, window.getOrDefault(ch, 0) + 1);

            if (need.containsKey(ch) &&
                window.get(ch).intValue() == need.get(ch).intValue()) {

                have++;
            }

            while (have == needCount) {

                if (right - left + 1 < minLength) {
                    minLength = right - left + 1;
                    answer = s.substring(left, right + 1);
                }

                char leftChar = s.charAt(left);

                window.put(
                    leftChar,
                    window.get(leftChar) - 1
                );

                if (need.containsKey(leftChar) &&
                    window.get(leftChar) < need.get(leftChar)) {

                    have--;
                }

                left++;
            }
        }

        return answer;
    }
}

File: DAY 4/solution.py/76. Minimum Window Substring.py
class Solution:
    def minWindow(self, s: str, t: str) -> str:

        if len(t) > len(s):
            return ""

        need = {}

        for ch in t:
            need[ch] = need.get(ch, 0) + 1

        window = {}

        left = 0
        have = 0
        needCount = len(need)

        minLength = float("inf")
        answer = ""

        for right in range(len(s)):

            ch = s[right]

            window[ch] = window.get(ch, 0) + 1

            if ch in need and window[ch] == need[ch]:
                have += 1

            while have == needCount:

                if right - left + 1 < minLength:
                    minLength = right - left + 1
                    answer = s[left:right + 1]

                leftChar = s[left]

                window[leftChar] -= 1

                if leftChar in need and window[leftChar] < need[leftChar]:
                    have -= 1

                left += 1

        return answer